### PR TITLE
Define loss function in Python

### PR DESCRIPTION
**Pull request by pluskid**
We already have the ability to define symbolic operators in Python. But due to multi-threading issue, this feature cause a lot of pain to implement in other languages like Julia (e.g. dmlc/MXNet.jl#16). With the new module system, we can achieve similar thing (at least for the case of simple linear chaining) by defining a loss module directly in Python.

This PR is a simple demo of defining a multiclass-hinge loss and train on MNIST.

Later on (???) I will try to port the module system to Julia.

### CODE REVIEW

**Comment by tqchen** [comment]
Looks nice, will this works naturally with the data parallel case?

**Comment by tqchen** [comment]
Also cc @jermainewang Maybe minpy could be used here as well ?

**Comment by pluskid** [comment]
@tqchen yes, the computation graph module is independent of the loss module. The first part can do data parallel as usual, the latter part can also do whatever they like. I guess for example people could implement the loss function with pycuda etc.

**Comment by pluskid** [comment]
I guess it might be possible to define a minpy module that wraps Amy minpy function and utilize the auto grad feature of minpy.

**Comment by tqchen** [comment]
Yes, that is what I mean. Could be interesting for a separate PR
